torch/csrc/autograd/engine.cpp

```diff
@@ -330,24 +357,44 @@ void Engine::reentrant_thread_init() {
     ++thread_pool_shared_->num_workers_;
     tp_shared->work_.wait(lk, [&tp_shared]{ return !tp_shared->graphtasks_queue_.empty();});
     --thread_pool_shared_->num_workers_;
-    auto graph_task = tp_shared->graphtasks_queue_.front();
+    auto task = tp_shared->graphtasks_queue_.front();
     tp_shared->graphtasks_queue_.pop();
     lk.unlock();
+    std::shared_ptr<GraphTask> graph_task;
+    if (!(graph_task = task.lock())) {
+      LOG(INFO) << "GraphTask has expired, skipping reentrant execution";
+      continue;
+    }
     set_device(graph_task->owner_);
     total_depth = graph_task->reentrant_depth_;
-    thread_main(graph_task);
+    thread_main(graph_task, /* reentrant thread*/ true);
```

Comment:
On second thought, I don't think `thread_main` should keep a graph task alive. Consider the following situation: you have a thread which reentrantly calls backwards. You spawn a reentrant thread which should return once the sub-backwards call is run. But then this sub-backwards has an error in one of its nodes, causing the GraphTask to go dead. We should stop executing all of the other nodes and raise an error through Python land. But you won't do this because the reentrant thread is keeping the graph task live.

Reply:
If the reentrant thread hits an error it would set an appropriate error on the graph task and eventually when all work is done we would still report this error to the user. 

> We should stop executing all of the other nodes and raise an error through Python land

We do this only in the case of dist_engine since we set `exit_on_error` only in that case. In our current autograd engine, we would try to execute all other outstanding tasks and return an error to the user once all tasks have completed.

Reply:
> If the reentrant thread hits an error it would set an appropriate error on the graph task and eventually when all work is done we would still report this error to the user.

This is true. But if we're not going to take care about not keeping graphs unnecessarily live, we might as well not use `weak_ptr` (and save ourselves some bugs).